### PR TITLE
switch the json backend to json-iterator

### PR DESCRIPTION
**Pull request by neolit123**
add a couple of json interator configurations:

1) jsonIterator should be compatible with the old encoding/json
there is a breaking change here because the "JSONOpt" is now
a no-op, although the options in encoding/json are not very useful.

2) caseSensitiveStrictJSONIterator is compatible with
econding/json WRT to DisallowUnknownFields, but it also has
CaseSensitive. This is a breaking change for strict unmarshal users
that previously allowed case-insensitive fields.

Something else to note is that json-iter does not seem to tolerate
field keys called "true", which the one from stdlib does.
Unit tests had to be adapted because of that.

Other changes:
- add a new public method UnmarshalWithConfig that allows passing
a json-iter configuration
- update/add unit tests
- update go.mod/sum
- remove yaml_go110*.go
These files had the purpose to handle DisallowUnknownFields for
json.Decoder which is no longer needed.

/kind feature
/priority important-longterm

fixes https://github.com/kubernetes-sigs/yaml/issues/17
fixes https://github.com/kubernetes-sigs/yaml/issues/15

/assign @dims 
/cc @sttts @errordeveloper @stealthybox
/hold

### CODE REVIEW

File: yaml.go [hunk 1]
```diff
@@ -4,17 +4,60 @@ import (
...
+// in some other library will mess with every usage of the jsoniter library in the whole program.
+// See https://github.com/json-iterator/go/issues/265
+var jsonIterator = createJSONIterator()
+var caseSensitiveStrictJSONIterator = createCaseSensitiveStrictJSONIterator()
+
 // Marshal marshals the object into JSON then converts JSON to YAML and returns the
 // YAML.
 func Marshal(o interface{}) ([]byte, error) {
-	j, err := json.Marshal(o)
+	j, err := caseSensitiveStrictJSONIterator.Marshal(o)
```
Comment (liggitt) [inline yaml.go]: @smarterclayton, didn't you see performance regressions in json-iter marshaling?

Comment (sttts) [reply yaml.go]: Would they really be relevant for YAML encoding? It's not that we do that in critical paths, do we?

**Comment by martina-if** [comment]
Thanks for making these changes! LGTM Is there anything we can do to help get this merged? :)

**Comment by errordeveloper** [comment]
Hey @neolit123, as Martina said - we'd be happy to help :)

File: yaml.go (hunk 1 above)
Comment (errordeveloper) [reply yaml.go]: Also, this doesn't affect all of YAML code paths in the current form.

**Comment by neolit123** [comment]
@errordeveloper up to the reviewers at this point. :)

File: yaml.go (hunk 1 above)
Comment (nikhita) [reply yaml.go]: Btw json-iterator isn't fully compatible with the standard library right now, created https://github.com/json-iterator/go/pull/371. PTAL.

Comment (nikhita) [reply yaml.go]: https://github.com/json-iterator/go/pull/371 has merged.

**Comment by nikhita** [comment]
> @smarterclayton, didn't you see performance regressions in json-iter marshaling?

Ref: https://github.com/kubernetes/kubernetes/pull/70574#issuecomment-476370725, https://github.com/kubernetes/kubernetes/pull/63242

The second PR referenced above did have high allocations:

Running `go test ./vendor/k8s.io/apiserver/pkg/endpoints/ -test.bench='Benchmark(Get|WatchHTTP)' -test.run=XXX -benchmem -test.memprofile=/tmp/mem2 -test.benchtime 5s`

On k/k master right now:

```
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints
BenchmarkGet-4              	   30000	    256834 ns/op	   15794 B/op	     127 allocs/op
BenchmarkWatchHTTP-4        	  100000	     92040 ns/op	    1523 B/op	      24 allocs/op
BenchmarkWatchHTTP_UTF8-4   	  100000	     96523 ns/op	    1829 B/op	      26 allocs/op
PASS
ok  	k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints	31.248s
```

k/k right now with changes in the second PR (https://github.com/kubernetes/kubernetes/pull/63242):
```
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints
BenchmarkGet-4                     10000	    691435 ns/op	   19122 B/op	     163 allocs/op
BenchmarkWatchHTTP-4        	  200000	     48193 ns/op	    5495 B/op	      64 allocs/op
BenchmarkWatchHTTP_UTF8-4   	  200000	     53738 ns/op	    5909 B/op	      69 allocs/op
PASS
ok  	k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints	28.604s
```

k/k with changes in this PR:

```
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints
BenchmarkGet-4              	   50000	    139204 ns/op	   15793 B/op	     127 allocs/op
BenchmarkWatchHTTP-4        	  200000	     56627 ns/op	    1522 B/op	      24 allocs/op
BenchmarkWatchHTTP_UTF8-4   	  100000	     62922 ns/op	    1829 B/op	      26 allocs/op
PASS
ok  	k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints	27.378s
```

So it should probably be fine...

**Comment by smarterclayton** [comment]
You should also run the serialization benchmarks in pkg/api/testing/serialization_test.go to get a before and after round trip time (that tests several variants of objects, whereas BenchmarkGet tests a single known object).

**Comment by nikhita** [comment]
> You should also run the serialization benchmarks in pkg/api/testing/serialization_test.go to get a before and after round trip time (that tests several variants of objects, whereas BenchmarkGet tests a single known object).

Running existing benchmarks (before and after) yielded similar results...but that isn't surprising because none of the benchmarks actually look for yaml marshaling and unmarshaling.

Created https://github.com/kubernetes/kubernetes/pull/78688 for adding benchmarks for yaml marshaling and unmarshaling as well. Using these benchmarks, incorporating changes in this PR into k/k yield higher allocations... :(

Running `go test k8s.io/kubernetes/pkg/api/testing/ -test.bench='BenchmarkEncodeYAMLMarshal|BenchmarkDecodeIntoYAML' -test.run=XXX -benchmem -test.memprofile=/tmp/mem2 -test.benchtime 5s`

**Before**:

```
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/api/testing
BenchmarkEncodeYAMLMarshal-4   	   10000	    781873 ns/op	  256102 B/op	    1438 allocs/op
BenchmarkDecodeIntoYAML-4      	   10000	    627891 ns/op	  167788 B/op	    1485 allocs/op
PASS
ok  	k8s.io/kubernetes/pkg/api/testing	14.354s
```

**After**:

```
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/api/testing
BenchmarkEncodeYAMLMarshal-4   	   10000	    867119 ns/op	  260776 B/op	    1443 allocs/op
BenchmarkDecodeIntoYAML-4      	   10000	    692040 ns/op	  181813 B/op	    1609 allocs/op
PASS
ok  	k8s.io/kubernetes/pkg/api/testing	15.842s
```

**Comment by smarterclayton** [comment]
What is the underlying goal of this PR?  Use json-iterator because <xxx>?  Consistency with Kube is important, but it's not all.  Was there a discussion about the why?  I don't see many comments in #17

**Comment by neolit123** [comment]
> What is the underlying goal of this PR? Use json-iterator because ? Consistency with Kube is important, but it's not all. Was there a discussion about the why? I don't see many comments in #17

main driver was discussed in https://github.com/kubernetes-sigs/yaml/issues/15
the current library's strict unmarshal is not case-sensitive.

**Comment by neolit123** [comment]
here is an example of this problem in the wild:
https://github.com/kubernetes/kubernetes/issues/80582

kubeadm has strict unmarshaling to show warnings and then perform regular unmarshall (two pass).
however the upper case `ClusterConfiguration` field was not caught; should have been `clusterConfiguration`.

**Comment by neolit123** [comment]
so going back to the OP and the discussion there are at lease 3 concerns here:

- benchmark results
the overhead after the change is not high and fixing the bug for StrictUnmarshal users can justify the small decrease in performance. on the other hand there aren't that many StrictUnmarshal users out there, so the decrease in performance is probably making a more apparent tilt of the scale.

- breaking change 1:
> This is a breaking change for strict unmarshal users
that previously allowed case-insensitive fields.

we enabled strict unmarshal in this library not so long ago, and given strict unmarshal is optional it seems that this might not affect that many users. but this is still probably going to break someone, somewhere once they move to the latest SHA of this repository.

- breaking change 2:
> there is a breaking change here because the "JSONOpt" is now
a no-op, although the options in encoding/json are not very useful.

not sure if this will break anyone, probably not.

---
i will update this PR and leave it for a couple of more weeks.

**Comment by neolit123** [comment]
clearing some pending PRs from my backlog.
if someone feels like they want to take this over i will leave the branch.
thanks!

**Comment by neolit123** [comment]
we are seeing more cases of this in the wild:
https://github.com/kubernetes/kubeadm/issues/2167

**Pull request closed without merging**